android/src/com/mapswithme/util/HttpClient.java

```diff
@@ -0,0 +1,246 @@
+        if (TextUtils.isEmpty(p.contentType))
+          throw new NullPointerException("Please set Content-Type for request.");
+
+        // Work-around for situation when more than one consequent POST requests can lead to stable
+        // "java.net.ProtocolException: Unexpected status line:" on a client and Nginx HTTP 499 errors.
+        // The only found reference to this bug is http://stackoverflow.com/a/24303115/1209392
+        connection.setRequestProperty("Connection", "close");
+        connection.setRequestProperty("Content-Type", p.contentType);
+        if (!TextUtils.isEmpty(p.contentEncoding))
+          connection.setRequestProperty("Content-Encoding", p.contentEncoding);
+
+        connection.setDoOutput(true);
+        if (p.data != null)
+        {
+          connection.setFixedLengthStreamingMode(p.data.length);
+          final OutputStream os = connection.getOutputStream();
+          try
+          {
+            os.write(p.data);
+          }
+          finally
+          {
+            os.close();
+          }
+          if (p.debugMode)
+            Log.d(TAG, "Sent " + p.httpMethod + " with content of size " + p.data.length);
+        }
+        else
+        {
+          final File file = new File(p.inputFilePath);
+          connection.setFixedLengthStreamingMode((int) file.length());
+          final BufferedInputStream istream = new BufferedInputStream(new FileInputStream(file), STREAM_BUFFER_SIZE);
+          final BufferedOutputStream ostream = new BufferedOutputStream(connection.getOutputStream(), STREAM_BUFFER_SIZE);
+          final byte[] buffer = new byte[STREAM_BUFFER_SIZE];
+          int bytesRead;
+          while ((bytesRead = istream.read(buffer, 0, STREAM_BUFFER_SIZE)) > 0)
+          {
+            ostream.write(buffer, 0, bytesRead);
+          }
+          istream.close(); // IOException
+          ostream.close(); // IOException
+          if (p.debugMode)
+            Log.d(TAG, "Sent " + p.httpMethod + " with file of size " + file.length());
+        }
+      }
+      // GET data from the server or receive response body
+      p.httpResponseCode = connection.getResponseCode();
+      if (p.debugMode)
+        Log.d(TAG, "Received HTTP " + p.httpResponseCode + " from server.");
+
+      if (p.httpResponseCode >= 300 && p.httpResponseCode < 400)
+        p.receivedUrl = connection.getHeaderField("Location");
+      else
+        p.receivedUrl = connection.getURL().toString();
+
+      p.contentType = connection.getContentType();
+      p.contentEncoding = connection.getContentEncoding();
+      final Map<String, List<String>> headers = connection.getHeaderFields();
+      if (headers != null && headers.containsKey("Set-Cookie"))
+      {
+        p.cookies = "";
+        for (final String value : headers.get("Set-Cookie"))
+        {
+          // Multiple Set-Cookie headers are normalized in C++ code.
+          if (value != null)
+            p.cookies += value + ", ";
```

Comment:
FYI: Java compiler will substitute StringBuilder internally if it will consider it as good optimization, thus this concatenation shouldn't be a problem. It would be problem if this concatenation will be used in a loop, in this case compiler will not apply any optimization 

http://stackoverflow.com/questions/1532461/stringbuilder-vs-string-concatenation-in-tostring-in-java